Integrate categories into the call tree

mstange opened the pull request:
This implements some of #519.

Deploy preview with example profile: [Link](https://deploy-preview-1058--perf-html.netlify.com/public/fa8770cf44c0d2ca9fa9776bb9d583acffa8db07/calltree/?hiddenThreads=5&thread=2&threadOrder=0-2-3-4-5-6-1&v=3)

This changes the format of the processed stackTable so I need to add an upgrader.

mstange commented:
Ok, this is ready for review now. The commits are separated for ease of reviewing, but the order might be a bit surprising. I tried to reorder the commits but it was too much work because I touched the same code multiple times. So I recommend looking at the list of commits first so that you know what to expect.

I have added two basic tests for the call tree categories. There are more tests I could add. For example, I don't have a test for the case where you "merge" a label frame into its caller, and that frame's descendants want to keep the category they inherited from the merged-away frame.

julienw commented on src/types/profile.js:
```diff
@@ -20,9 +20,36 @@ export type resourceTypeEnum = number;
+ * prefix stack. Root stacks whose frame has a null stack have their category
+ * set to the "default category". (The default category is currently defined as
+ * the category in the profile's category list whose color is "grey", and such
+ * a category is required to be present.)
```
we should enforce this at processing time: if no such category exist, we should create it. (cf other comments about the same subject)

julienw commented on src/types/profile.js:
```diff
@@ -20,9 +20,36 @@ export type resourceTypeEnum = number;
+ * You could argue that the stack table's category column is derived data and as
+ * such doesn't need to be stored in the profile itself. This is true, but
+ * storing this information in the stack table makes it a lot easier to carry
+ * it through various transforms that we apply to threads.
```
The alternative is recomputing it from the transformed frame information. This would allow that we don't really mind about it, but we couldn't likely manage conflicts as easily (which I'm still not sure we should).  

I'll trust you here as you thought about this more than I did.

mstange replied:
I'll add a sentence to this comment that gives an example of where recomputing it from the transformed frame information would not be good enough.

mstange replied:
I'm adding the following sentence to the comment:

```
 * For example, here's a case where a stack's category is not recoverable from
 * any other information in the transformed thread:
 * In partial the call path
 *   someJSFunction [JS] -> Node.insertBefore [DOM] -> nsAttrAndChildArray::InsertChildAt,
 * the stack node for nsAttrAndChildArray::InsertChildAt should inherit the
 * category DOM from its "Node.insertBefore" prefix stack. And it should keep
 * the DOM category even if you apply the "Merge node into calling function"
 * transform to Node.insertBefore. This transform removes the stack node
 * "Node.insertBefore" from the stackTable, so the information about the DOM
 * category would be lost if it wasn't inherited into the
 * nsAttrAndChildArray::InsertChildAt stack before transforms are applied.
```

julienw commented on src/profile-logic/processed-profile-versioning.js:
```diff
@@ -595,5 +595,35 @@ const _upgraders = {
+
+    for (const thread of threads) {
+      const { stackTable, frameTable } = thread;
+      stackTable.category = [];
```
same comment as another one earlier: because we know what the final length is, we could as well specify it already.

julienw commented on src/types/profile.js:
```diff
@@ -15,17 +15,45 @@ export type IndexIntoStringTable = number;
  */
 export type StackTable = {
   frame: IndexIntoFrameTable[],
+  category: IndexIntoCategoryList[],
```
Maybe a newbie question: why are we using normal arrays here, but Int32Arrays in the CallNodeTable ?

mstange replied:
No good reason. I think when I created the call node table, I had to work really hard to get it to perform well, so I tried to squeeze out performance in every corner that I could find. But I don't think I've ever properly benchmarked the typed array solution versus the regular array solution.

It's possible that the performance of the stackTable would also benefit from typed arrays.

There's one difference between the callNodeTable and the tables in the profile that makes it easier to use typed arrays in the callNodeTable: The callNodeTable is never stored inside the profile - it's just external (derived) information that's used for display. So we don't need to worry about serializing it. But with data in the profile, we need a way of serializing it to JSON / deserializing it from JSON. We already have some custom code for the stringTable, so it wouldn't be too hard to add more custom serialization code for the other tables, but I'm not sure if it's worth doing.

julienw commented on src/test/fixtures/profiles/make-profile.js:
```diff
@@ -294,7 +312,8 @@ function _parseTextSamples(textSamples: string): Array<string[]> {
 function _buildThreadFromTextOnlyStacks(
   textOnlyStacks: Array<string[]>,
-  funcNames: string[]
+  funcs: Array<{| funcName: string, category: number | null |}>,
```
nit: Instead of number, should it be IndexToCategoryList ?

mstange replied:
Yes it should. But this comment no longer applies because I've changed the argument back to `funcNames` during the rebase, so there's nothing to address here.

julienw commented on src/profile-logic/profile-data.js:
```diff
@@ -269,6 +292,11 @@ function _filterThreadByFunc(
+          } else if (
+            newStackTable.category[newStack] !== stackTable.category[stackIndex]
+          ) {
+            newStackTable.category[newStack] = defaultCategory;
```
I put the comment here but this applies to all other code changes about setting a default category when there's a conflict: I'm not so sure about this. I think that in most (if not all) of the cases, one of the conflicting categories should be kept instead of using the default category.
_or_ we should somehow keep the information of all categories.

With this last "or" I've also in mind the information we want to display in the sidebar: ideally I'd like to display all "collapsed" categories if I select a node resulting from the collapse.

But I believe this is more work than what we want to do now, so let's focus on the first proposal instead: that one of the categories should "win" over the others.

julienw replied:
\+ please add some comments at each location

mstange replied:
I'm not convinced that either of the two solutions you presented is what we want.

 - Keeping around all individual categories: This would mean that, per stack, you'd need to keep around a dynamic list of categories. And for each category, you'd probably also want to store the number of samples that were in a stack of that category. And then you'd need to update those sample counts as the range selection changes, for example.

 - Picking one of the collapsed categories instead of a "neutral" category: I don't like this idea because it "lies" without acknowledging the lie. Which category would you pick? The first one that you encounter in the stackTable? Or the one with the most samples? If the latter, do you adjust your pick when the range selection changes?

Picking a neutral category acknowledges that some information has been lost. Here are two screenshots of it in action, from [this profile](https://perfht.ml/2Lkr50u):

 1. With an inverted call tree. General functions like waiting on a lock or memmove can be called from stacks of wildly different categories and it doesn't really make sense to assign a single category to them.
<img width="1136" alt="screen shot 2018-07-17 at 2 28 57 pm" src="https://user-images.githubusercontent.com/961291/42838132-ee7a11d0-89cd-11e8-97ce-a7b37e047995.png">

 2. With a non-inverted call tree, after a "focus function" transform on a hashtable search. Hashtable searching is also a very general operation that does not belong to a single category.
<img width="982" alt="screen shot 2018-07-17 at 2 08 10 pm" src="https://user-images.githubusercontent.com/961291/42838804-c5b8267c-89cf-11e8-889d-1b12ba91761d.png">

You can see that in the hashtable searching example, most of the call nodes under PLDHashTable::Search still have a non-default category, because these functions happened to be sampled only while they were in a stack of a particular category. This might be a bit misleading - we might want to reset them to the default category, too. But there are probably cases where showing the original category is actually useful. So I'd say we keep this behavior for now and see how well it works.

A word about the sidebar: I believe in the sidebar we'll want to display a breakdown of the categories for the *samples* that ran under the selected call tree node. We can recover this information from the unfiltered thread, as follows:

 1. For the selected call tree node in the filtered thread, enumerate all samples that match this call tree node.
 2. Get the *unfiltered* thread and look up the *unfiltered* stacks for these samples.
 3. From the unfiltered stacks, look up the category for each sample's leaf stack.

These categories are your original sample categories, and form your sidebar category breakdown.

I took this approach in the thread activity graph: The thread activity graph component has two thread props: `fullThread` and `filteredThread`. It draws category colors based on `fullThread` and makes call tree decisions based on `filteredThread`, under the assumption that the `samples` tables from both threads are compatible.

julienw replied:
Thanks for the thoughtful comment. I'll have to think more about the end of the comment about the sidebar.

julienw commented on res/css/style.css:
```diff
@@ -193,6 +193,43 @@ body, #root, .profileViewer {
 }
 
+.treeViewCategoryKnob {
+  display: inline-block;
```
as an inline-block, I find it's slightly too high compared to the text and the expanding arrow. Maybe it needs `vertical-align: middle` ? I tried with this and it's not satisfying either... so maybe don't do anything :)

also some of them are too small on my computer, depending on the zoom level, see:
![screenshot](https://user-images.githubusercontent.com/454175/41107762-4595b7c8-6a73-11e8-8053-a6e32b5467bb.png)
The purple one misses some height at some zoom levels...
But let's not spend too much time on this if you don't have a good idea to fix it.

mstange replied:
Ensuring that something is square at a non-integer devicePixelRatio is basically impossible if you don't want to resort to an image. I'm not going to try to fix this.

mstange commented on src/profile-logic/process-profile.js:
```diff
@@ -507,9 +508,34 @@ function _processFrameTable(
+    } else {
+      const prefix = geckoStackTable.prefix[stackIndex];
+      if (prefix !== null) {
+        stackCategory = categoryColumn[prefix];
```
Yes it is. I'll add a comment.

mstange commented on src/profile-logic/process-profile.js:
```diff
@@ -23,6 +23,7 @@ import type {
   Profile,
```
The result of the "old cleopatra format" converter is a processed profile of processed profile format version zero, which then runs through the processed profile version upgraders. The idea was never having to tweak `old-cleopatra-profile-format.js` ever again.

julienw replied:
ah right, I forgot about this.

mstange commented on src/test/fixtures/profiles/make-profile.js:
```diff
@@ -235,56 +255,22 @@ function _parseTextSamples(textSamples: string): Array<string[]> {
+  const rows = lines.map(line =>
+    columnPositions.map((pos, columnIndex) =>
+      line.substring(pos, columnPositions[columnIndex + 1]).trim()
+    )
```
"same number of columns in each line" isn't required for a well-formed input: different samples can have stacks of different heights.

julienw replied:
ah right, I knew I missed something !

mstange commented on src/test/fixtures/profiles/make-profile.js:
```diff
@@ -144,18 +144,22 @@ export function getEmptyThread(overrides: ?Object): Thread {
+ *   B.js       B.js      F [Graphics]  F
+ *   C.js       C.js      G             G
+ *   D          D                       H [DOM]
+ *   E [Other]  E [Other]
```
I was going back and forth on this myself. One reason I did it was that we have a category which has spaces in the category name itself: "GC / CC". The other reason I wanted to do it is because it allows more extensions to this format without having to squish them all together; for example, we might want to add ways to have a func with multiple frames (where each of the frame has a different frame address), or we might want to explicitly specify the JS implementation on a frame.

julienw replied:
I see, makes sense, thanks !

julienw commented on src/test/fixtures/profiles/make-profile.js:
```diff
@@ -144,18 +144,22 @@ export function getEmptyThread(overrides: ?Object): Thread {
+ * Functions ending in "js" are marked as JS functions in the funcTable's isJS
+ * column. Functions that have a category name in square brackets after them
+ * are annotated with that category in the frameTable's category column.
+ *
```
A quick note because I changed this function in https://github.com/devtools-html/perf.html/pull/1041/files#diff-f74cb1c027c114cfd23055a90af53ee5 to be able to specify optimizations after all. I used a similar solution to yours by using backets, but still sticked to the function name, and with a prefix, eg: `Cjs[jit:ion]`.

mstange commented on src/profile-logic/process-profile.js:
```diff
@@ -507,9 +508,34 @@ function _processFrameTable(
+): StackTable {
+  // Compute a non-null category for every stack
+  const greyCategory = categories.findIndex(c => c.color === 'grey') || 0;
+  const categoryColumn = [];
```
This way of creating arrays does not pre-allocate space, it only sets `length` to a different value. I've found [a benchmark](https://jsperf.com/array-growth-dynamic-vs-preset) that compares the two; both approaches have the same performance and grow the buffer dynamically.

julienw replied:
so I actually got a discussion with @nbp about this. If you set the length, it will preallocate, but with a "max" boundary (around 1200 slots maximum IIRC).
I don't know what the behavior is when resizing: is it resizing with bigger chunks when it then resizes ? Maybe @nbp can help answering this question.

Also the benchmark is slightly different in that you're using `push` in this code. This may trigger different behavior.

That said we can also look at this again later if we find there's a perf issue.

nbp replied:
Indeed using `new Array(nb)` will reserve space up to `nb` for small numbers.  Using `push` will amortize the allocation cost by doubling the size.

mstange commented on src/profile-logic/transforms.js:
```diff
@@ -843,6 +851,7 @@ export function collapseResource(
           // Add the new stack.
           newStackTable.prefix.push(newStackPrefix);
           newStackTable.frame.push(collapsedFrameIndex);
+          newStackTable.category.push(category);
```
Yes I did, thanks for catching that bug! I actually noticed this bug while I was using the deploy preview, but I had forgotten about your comment, so I'm glad I read through your review again before starting to debug it :)

mstange commented on src/profile-logic/call-tree.js:
```diff
@@ -436,6 +442,7 @@ export function getCallTree(
   callNodeInfo: CallNodeInfo,
+  categories: CategoryList,
   implementationFilter: string,
   invertCallstack: boolean
```
I started doing this but wasn't happy with it, so I reverted it again. `getCallTree` is called in many places, and passing `profile.meta` everywhere looked more confusing, I thought.

julienw replied:
:+1: thx for trying !

mstange commented on src/test/fixtures/profiles/make-profile.js:
```diff
@@ -212,6 +213,25 @@ export function getProfileFromTextSamples(
+  const ranges = [];
+  let match;
+  while ((match = regex.exec(str)) !== null) {
+    ranges.push({ start: match.index, end: match.index + match[0].length });
```
I think it's less intuitive, so I'm going to keep what I have.

julienw replied:
:+1:

julienw commented on src/test/fixtures/profiles/make-profile.js:
```diff
@@ -184,35 +188,58 @@ export function getProfileFromTextSamples(
+  str,
+  categories
+): {| funcName: string, category: number | null |} {
+  const match = /(.*) \[(.+)\]/.exec(str);
```
:+1:

mstange commented:
I've pushed the new version. Most review comments have been addressed in follow-up commits; the original commits have only been updated as far as it was necessary for rebasing.

> I have some questions, but not all of them are blocking the patch:
> 
> - About the fact we put the category in the stackTable. You mentioned it yourselves in a comment. I'm not totally convinced, but that's not something we can't change later, so I'm fine keeping it that way for now, unless you reconsider it.

I've addressed this question with the comment that I added:

```
 * You could argue that the stack table's category column is derived data and as
 * such doesn't need to be stored in the profile itself. This is true, but
 * storing this information in the stack table makes it a lot easier to carry
 * it through various transforms that we apply to threads.
 * For example, here's a case where a stack's category is not recoverable from
 * any other information in the transformed thread:
 * In partial the call path
 *   someJSFunction [JS] -> Node.insertBefore [DOM] -> nsAttrAndChildArray::InsertChildAt,
 * the stack node for nsAttrAndChildArray::InsertChildAt should inherit the
 * category DOM from its "Node.insertBefore" prefix stack. And it should keep
 * the DOM category even if you apply the "Merge node into calling function"
 * transform to Node.insertBefore. This transform removes the stack node
 * "Node.insertBefore" from the stackTable, so the information about the DOM
 * category would be lost if it wasn't inherited into the
 * nsAttrAndChildArray::InsertChildAt stack before transforms are applied.
```

> - conflicting categories in transforms / filter / inverted call tree. I think this makes sense for some of them (inverted call tree) and less for others (filter by implementation). Again, this is something we can easily reconsider later, so it's not blocking.

I've addressed this question an another thread on this issue.

> - How the default category is defined: I don't like this empirical way of finding it by relying on its color. I don't want this to block the patch but I'd like very much to get your opinion on this concern.

I agree, this is pretty bad and I haven't fixed it. I want to create a new PR to fix this properly:

 - Give the processed profile a field `profile.meta.defaultCategory`
 - write an upgrader that sets it to the grey category, and adds it to the category list if not present
 - do the same thing during profile processing

> - How we specify the colors: I thought it wouldn't be an hardcoded list but would be much more free. This is something we can evolve later so it's not blocking this patch.

The reason I made it a hardcoded list is so that we can make minor adjustments of the colors in the frontend without having to change the profile JSON generation, and so that other producers of profiles don't have to pick precise RGB values.
I haven't documented the list anywhere yet. This still remains to be done.

julienw commented on src/types/profile.js:
```diff
@@ -15,17 +15,56 @@ export type IndexIntoStringTable = number;
+ * it through various transforms that we apply to threads.
+ * For example, here's a case where a stack's category is not recoverable from
+ * any other information in the transformed thread:
+ * In partial the call path
```
nit: I don't understand this "In partial". Should this be something like `When parsing` ?

mstange replied:
Oops, this was supposed to say "In *the* partial call path [...]". I'm going to change it back to "In the call path [...]" which is what I had there first.
(I added the word "partial" because that path, as written, doesn't actually occur in real Gecko profiles because I left out a few functions in between.)
